### PR TITLE
Worldtube component

### PR DESCRIPTION
**Pull request by moxcodes**
## Proposed changes
Add worldtube h5 component and initializer. The remaining actions to interface between components will have to wait until after the PRs introducing the other CCE singletons. 

### Types of changes:

- [ ] Bugfix
- [x] New feature
- [ ] Refactor

### Component:

- [x] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests and `clang-tidy`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### Dependencies
- [x] #1869 
- [x] #1892

### CODE REVIEW

File: src/Evolution/Systems/Cce/Actions/InitializeWorldtubeBoundary.hpp [hunk 1]
```diff
@@ -0,0 +1,99 @@
+// Distributed under the MIT License.
```
Comment (nilsdeppe) [inline src/Evolution/Systems/Cce/Actions/InitializeWorldtubeBoundary.hpp:1]: Can you factor the changes in this commit into several smaller commits? There appear to be a lot of fairly independent changes.

**Review by nilsdeppe — CHANGES_REQUESTED** [review]

**Comment by moxcodes** [comment]
split into several commits.
Cheers!

File: src/Evolution/Systems/Cce/Components/WorldtubeBoundary.hpp [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+#include "Parallel/Actions/TerminatePhase.hpp"
+#include "Parallel/ConstGlobalCache.hpp"
+#include "Parallel/Info.hpp"
+#include "Parallel/Invoke.hpp"
+#include "ParallelAlgorithms/Initialization/Actions/RemoveOptionsAndTerminatePhase.hpp"
+
+namespace Cce {
+
+/*!
+ * \brief ParallelComponent for the management and reporting of CCE worldtube
```
Comment (nilsdeppe) [inline src/Evolution/Systems/Cce/Components/WorldtubeBoundary.hpp]: Component that supplies...

File: src/Evolution/Systems/Cce/Components/WorldtubeBoundary.hpp [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+ * boundary data.
+ *
+ * \details The \ref DataBoxGroup associated with the worldtube boundary
```
Comment (nilsdeppe) [inline src/Evolution/Systems/Cce/Components/WorldtubeBoundary.hpp:19]: Can we split the first sentence into multiple?

File: src/Evolution/Systems/Cce/Components/WorldtubeBoundary.hpp [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+ * component will contain a data manager (e.g. `WorldtubeDataManager`) linked to
+ * an H5 file, and will handle buffering and interpolating to desired target
+ * time points when requested via the simple action
+ * `BoundaryComputeAndSendToExtractor`, at which point it will send the required
+ * collection of boundary quantities to the identified 'Extractor' component. It
+ * is assumed that the simple action `BoundaryComputeAndSendToExtractor` will
```
Comment (nilsdeppe) [inline src/Evolution/Systems/Cce/Components/WorldtubeBoundary.hpp]: `SendToExtractor` -> `SendToEvolution`

File: src/Evolution/Systems/Cce/Components/WorldtubeBoundary.hpp [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+ * only be called during the `Extraction` phase.
```
Comment (nilsdeppe) [inline src/Evolution/Systems/Cce/Components/WorldtubeBoundary.hpp]: `Extraction` -> `Evolution` (or `Evolve`, whatever is currently done in the execs)

File: src/Evolution/Systems/Cce/Tags.hpp [hunk 6]
```diff
@@ -298,5 +302,13 @@ struct BondiR : db::SimpleTag {
   using type = Scalar<SpinWeighted<ComplexDataVector, 0>>;
   static std::string name() noexcept { return "R"; }
 };
+
+/// A simple tag for the `WorldtubeDataManager`
+struct H5WorldtubeBoundaryDataManager : db::SimpleTag {
+  using type = WorldtubeDataManager;
+  static std::string name() noexcept {
```
Comment (nilsdeppe) [inline src/Evolution/Systems/Cce/Tags.hpp]: can remove now

File: src/Evolution/Systems/Cce/Tags.hpp [hunk 7]
```diff
@@ -298,5 +302,13 @@ struct BondiR : db::SimpleTag {
...
+    return "H5WorldtubeBoundaryDataManager";
+  }
+};
```
Comment (nilsdeppe) [inline src/Evolution/Systems/Cce/Tags.hpp:312]: Can you add a test that checks the name and sticks one of these into a DataBox and gets it back out?

File: src/Evolution/Systems/Cce/OptionTags.hpp [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+
+namespace Cce {
+/// \cond
+class Interpolator;
+/// \endcond
+
+namespace OptionTags {
+
+struct LMax {
+  using type = size_t;
```
Comment (nilsdeppe) [inline src/Evolution/Systems/Cce/OptionTags.hpp:25]: cstddef

File: src/Evolution/Systems/Cce/OptionTags.hpp [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+  static constexpr OptionString help{
+      "maximum l value for spin-weighted spherical harmonics"};
```
Comment (nilsdeppe) [inline src/Evolution/Systems/Cce/OptionTags.hpp]: `Maximum` repeat

File: src/Evolution/Systems/Cce/OptionTags.hpp [hunk 10]
```diff
@@ -0,0 +1,103 @@
...
+  using type = size_t;
+  static constexpr OptionString help{
+      "Number of radial grid points in the spherical domain"};
+};
+
+struct EndTime {
+  using type = double;
+  static constexpr OptionString help{"End time for the Cce Evolution."};
+  static double default_value() noexcept {
+    return std::numeric_limits<double>::quiet_NaN();
```
Comment (nilsdeppe) [inline src/Evolution/Systems/Cce/OptionTags.hpp]: `quiet_NaN` -> `infinity`

File: src/Evolution/Systems/Cce/OptionTags.hpp [hunk 11]
```diff
@@ -0,0 +1,103 @@
...
+  }
+};
+
+struct StartTime {
+  using type = double;
```
Comment (nilsdeppe) [inline src/Evolution/Systems/Cce/OptionTags.hpp:54]: Can you stick all the option tags into an option group for CCE?

"Your double offends me." - @fmahebert 2012

File: src/Evolution/Systems/Cce/OptionTags.hpp [hunk 12]
```diff
@@ -0,0 +1,103 @@
...
+
+struct TargetStepSize {
+  using type = double;
+  static constexpr OptionString help{"Target time step size for Cce Evolution"};
+};
+
+struct BoundaryDataFilename {
+  using type = std::string;
+  static constexpr OptionString help{
+      "h5 file to read the wordltube data from."};
```
Comment (nilsdeppe) [inline src/Evolution/Systems/Cce/OptionTags.hpp]: h5 -> H5

File: src/Evolution/Systems/Cce/OptionTags.hpp [hunk 13]
```diff
@@ -0,0 +1,103 @@
...
+  using type = double;
+  static constexpr OptionString help{"End time for the Cce Evolution."};
+  static double default_value() noexcept {
+    return std::numeric_limits<double>::quiet_NaN();
+  }
+};
+
+struct StartTime {
+  using type = double;
+  static constexpr OptionString help{"Start time for the Cce Evolution."};
```
Comment (nilsdeppe) [inline src/Evolution/Systems/Cce/OptionTags.hpp]: Make default some standard value (-inf or quiet NaN), in which case it chooses a "reasonable" start time (and document what that means)

File: src/Evolution/Systems/Cce/OptionTags.hpp [hunk 14]
```diff
@@ -0,0 +1,103 @@
...
+  static constexpr OptionString help{"Target time step size for Cce Evolution"};
+};
+
+struct BoundaryDataFilename {
+  using type = std::string;
+  static constexpr OptionString help{
+      "h5 file to read the wordltube data from."};
+};
+
+struct H5LookaheadPoints {
```
Comment (nilsdeppe) [inline src/Evolution/Systems/Cce/OptionTags.hpp]: Points -> Times

File: src/Evolution/Systems/Cce/OptionTags.hpp [hunk 15]
```diff
@@ -0,0 +1,103 @@
...
+  using type = size_t;
+  static constexpr OptionString help{
+      "Number of points to cache on each file read operation."};
```
Comment (nilsdeppe) [inline src/Evolution/Systems/Cce/OptionTags.hpp]: Number of _____ to cache
(time steps from H5 file)

File: src/Evolution/Systems/Cce/OptionTags.hpp [hunk 16]
```diff
@@ -0,0 +1,103 @@
...
+  static size_t default_value() noexcept { return 200; }
+};
+
+struct H5Interpolator {
+  static std::string name() noexcept { return "H5Interpolator"; }
```
Comment (nilsdeppe) [inline src/Evolution/Systems/Cce/OptionTags.hpp]: no name()

File: src/Evolution/Systems/Cce/OptionTags.hpp [hunk 17]
```diff
@@ -0,0 +1,103 @@
...
+  using type = std::unique_ptr<Interpolator>;
+  static constexpr OptionString help{
+      "The interpolator for imported h5 worldtube data."};
+};
+
+struct ScriInterpolationPoints {
+  static std::string name() noexcept { return "ScriInterpPoints"; }
+  using type = size_t;
+  static constexpr OptionString help{
+      "Number of points to interpolate for at scri+."};
```
Comment (nilsdeppe) [inline src/Evolution/Systems/Cce/OptionTags.hpp]: `Order of time interpolation at scri+`

File: src/Evolution/Systems/Cce/OptionTags.hpp [hunk 18]
```diff
@@ -0,0 +1,103 @@
...
+};
+
+struct H5Interpolator {
+  static std::string name() noexcept { return "H5Interpolator"; }
+  using type = std::unique_ptr<Interpolator>;
+  static constexpr OptionString help{
+      "The interpolator for imported h5 worldtube data."};
+};
+
+struct ScriInterpolationPoints {
```
Comment (nilsdeppe) [inline src/Evolution/Systems/Cce/OptionTags.hpp]: Points -> Order

File: src/Evolution/Systems/Cce/OptionTags.hpp [hunk 19]
```diff
@@ -0,0 +1,103 @@
...
+  static std::string name() noexcept { return "ScriInterpPoints"; }
```
Comment (nilsdeppe) [inline src/Evolution/Systems/Cce/OptionTags.hpp]: Again

File: src/Evolution/Systems/Cce/OptionTags.hpp [hunk 20]
```diff
@@ -0,0 +1,103 @@
...
+
+  static WorldtubeDataManager create_from_options(
+      const size_t l_max, const std::string& filename,
+      const size_t number_of_lookahead_points,
+      const std::unique_ptr<intrp::SpanInterpolator>& interpolator) noexcept {
+    return WorldtubeDataManager{
+      std::make_unique<SpecWorldtubeH5BufferUpdater>(filename), l_max,
+          number_of_lookahead_points, interpolator->get_clone()};
+  }
+  static std::string name() noexcept {
```
Comment (nilsdeppe) [inline src/Evolution/Systems/Cce/OptionTags.hpp]: Can you remove this?

File: src/Evolution/Systems/Cce/OptionTags.hpp [hunk 21]
```diff
@@ -0,0 +1,103 @@
...
+    return "H5WorldtubeBoundaryDataManager";
+  }
+};
```
Comment (nilsdeppe) [inline src/Evolution/Systems/Cce/OptionTags.hpp:116]: Can you add tests for the tags?

File: src/Evolution/Systems/Cce/Actions/InitializeWorldtubeBoundary.hpp [hunk 22]
```diff
@@ -0,0 +1,99 @@
...
+#include "Parallel/Info.hpp"
+#include "Parallel/Invoke.hpp"
+#include "ParallelAlgorithms/Initialization/MergeIntoDataBox.hpp"
+
+namespace Cce {
+/*!
+ * \ingroup ActionsGroup
+ * \brief Initializes a H5WorldtubeBoundary
+ *
+ * \details Uses: initialization tag
```
Comment (nilsdeppe) [inline src/Evolution/Systems/Cce/Actions/InitializeWorldtubeBoundary.hpp]: Can you make this a bulleted list?

File: src/Evolution/Systems/Cce/Actions/InitializeWorldtubeBoundary.hpp [hunk 23]
```diff
@@ -0,0 +1,99 @@
...
+
+  template <class Metavariables>
+  using h5_boundary_manager_simple_tags = db::AddSimpleTags<
+      ::Tags::Variables<
+          typename Metavariables::cce_boundary_communication_tags>,
+      Tags::H5WorldtubeBoundaryDataManager>;
+
+  template <class Metavariables>
+  using return_tag_list =
+      tmpl::append<h5_boundary_manager_simple_tags<Metavariables>>;
```
Comment (nilsdeppe) [inline src/Evolution/Systems/Cce/Actions/InitializeWorldtubeBoundary.hpp]: you can remove the append here

File: src/Evolution/Systems/Cce/Actions/InitializeWorldtubeBoundary.hpp [hunk 24]
```diff
@@ -0,0 +1,99 @@
...
+            Requires<tmpl::list_contains_v<
+                DbTags, InitializationTags::H5WorldtubeBoundaryDataManager>> =
+                nullptr>
+  static auto apply(db::DataBox<DbTags>& box,
+                    const tuples::TaggedTuple<InboxTags...>& /*inboxes*/,
+                    Parallel::ConstGlobalCache<Metavariables>& cache,
+                    const ArrayIndex& /*array_index*/,
+                    const ActionList /*meta*/,
+                    const ParallelComponent* const /*meta*/) noexcept {
+    const auto& l_max = Parallel::get<Spectral::Swsh::Tags::LMax>(cache);
```
Comment (nilsdeppe) [inline src/Evolution/Systems/Cce/Actions/InitializeWorldtubeBoundary.hpp]: `const size_t`

File: src/Evolution/Systems/Cce/Actions/InitializeWorldtubeBoundary.hpp [hunk 25]
```diff
@@ -0,0 +1,99 @@
...
+    Variables<typename Metavariables::cce_boundary_communication_tags>
+        boundary_variables{
+            Spectral::Swsh::number_of_swsh_collocation_points(l_max)};
+
+    WorldtubeDataManager boundary_data_manager;
+    db::mutate<InitializationTags::H5WorldtubeBoundaryDataManager>(
+        make_not_null(&box),
+        [&boundary_data_manager](const gsl::not_null<WorldtubeDataManager*>
+                                     initialization_data_manager) {
```
Comment (nilsdeppe) [inline src/Evolution/Systems/Cce/Actions/InitializeWorldtubeBoundary.hpp]: noexcept

File: src/Evolution/Systems/Cce/Actions/InitializeWorldtubeBoundary.hpp [hunk 26]
```diff
@@ -0,0 +1,99 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#pragma once
+
```
Comment (nilsdeppe) [inline src/Evolution/Systems/Cce/Actions/InitializeWorldtubeBoundary.hpp:5]: tuple

File: src/Evolution/Systems/Cce/Actions/InitializeWorldtubeBoundary.hpp [hunk 27]
```diff
@@ -0,0 +1,99 @@
...
+#include "DataStructures/DataBox/DataBoxTag.hpp"
+#include "Evolution/Systems/Cce/BoundaryData.hpp"
+#include "Evolution/Systems/Cce/OptionTags.hpp"
+#include "Evolution/Systems/Cce/ReadBoundaryDataH5.hpp"
+#include "Evolution/Systems/Cce/Tags.hpp"
+#include "NumericalAlgorithms/Interpolation/SpanInterpolator.hpp"
+#include "Parallel/Algorithm.hpp"
+#include "Parallel/Info.hpp"
+#include "Parallel/Invoke.hpp"
+#include "ParallelAlgorithms/Initialization/MergeIntoDataBox.hpp"
```
Comment (nilsdeppe) [inline src/Evolution/Systems/Cce/Actions/InitializeWorldtubeBoundary.hpp:20]: tmpl, requires, tagged tuple, gsl, DataBox.hpp, utility (for std::move), cstddef

File: tests/Unit/Evolution/Systems/Cce/Actions/Test_InitializeWorldtubeBoundary.cpp [hunk 28]
```diff
@@ -0,0 +1,87 @@
...
+#include "tests/Unit/Evolution/Systems/Cce/Actions/WorldtubeBoundaryMocking.hpp"
+#include "tests/Unit/Evolution/Systems/Cce/BoundaryTestHelpers.hpp"
+#include "tests/Unit/TestHelpers.hpp"
+#include "tests/Utilities/MakeWithRandomValues.hpp"
+
+namespace Cce {
+
+SPECTRE_TEST_CASE(
+    "Unit.Evolution.Systems.Cce.Actions.InitializeWorldtubeBoundary",
+    "[Unit]") {
```
Comment (nilsdeppe) [inline tests/Unit/Evolution/Systems/Cce/Actions/Test_InitializeWorldtubeBoundary.cpp]: Is there a `Cce` thingy for the tests? e.g. `[Unit][Cce]`? We could add one

File: tests/Unit/Evolution/Systems/Cce/Actions/Test_InitializeWorldtubeBoundary.cpp [hunk 29]
```diff
@@ -0,0 +1,87 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#include "tests/Unit/TestingFramework.hpp"
+
+#include "Evolution/Systems/Cce/Actions/InitializeWorldtubeBoundary.hpp"
+#include "Evolution/Systems/Cce/BoundaryData.hpp"
+#include "Evolution/Systems/Cce/Components/WorldtubeBoundary.hpp"
+#include "NumericalAlgorithms/Interpolation/BarycentricRationalSpanInterpolator.hpp"
+#include "PointwiseFunctions/AnalyticSolutions/GeneralRelativity/KerrSchild.hpp"
```
Comment (nilsdeppe) [inline tests/Unit/Evolution/Systems/Cce/Actions/Test_InitializeWorldtubeBoundary.cpp:20]: I'll let you figure out the includes :D

File: tests/Unit/Evolution/Systems/Cce/Actions/WorldtubeBoundaryMocking.hpp [hunk 30]
```diff
@@ -0,0 +1,45 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#pragma once
+
+#include "Evolution/Systems/Cce/Actions/InitializeWorldtubeBoundary.hpp"
+#include "Evolution/Systems/Cce/BoundaryData.hpp"
+#include "Evolution/Systems/Cce/Components/WorldtubeBoundary.hpp"
+#include "tests/Unit/ActionTesting.hpp"
```
Comment (nilsdeppe) [inline tests/Unit/Evolution/Systems/Cce/Actions/WorldtubeBoundaryMocking.hpp:15]: includes

**Review by nilsdeppe — CHANGES_REQUESTED** [review]
Group review with @kidder @fmahebert @moxcodes @markscheel @geoffrey4444

**Comment by moxcodes** [comment]
fixups posted.
Cheers!

File: tests/Unit/Evolution/Systems/Cce/Actions/Test_InitializeCharacteristicExtractor.cpp [hunk 31]
```diff
@@ -0,0 +1,268 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#include "tests/Unit/TestingFramework.hpp"
+
+#include "Evolution/Systems/Cce/Actions/InitializeWorldtubeBoundary.hpp"
```
Comment (kidder) [inline tests/Unit/Evolution/Systems/Cce/Actions/Test_InitializeCharacteristicExtractor.cpp]: fix the includes

File: tests/Unit/Evolution/Systems/Cce/Test_OptionTags.cpp [hunk 32]
```diff
@@ -0,0 +1,55 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#include "tests/Unit/TestingFramework.hpp"
+
+#include <string>
```
Comment (kidder) [inline tests/Unit/Evolution/Systems/Cce/Test_OptionTags.cpp]: need cstddef, check includes

File: tests/Unit/Evolution/Systems/Cce/Test_OptionTags.cpp [hunk 33]
```diff
@@ -0,0 +1,55 @@
...
+#include "Evolution/Systems/Cce/ReadBoundaryDataH5.hpp"
+#include "NumericalAlgorithms/Interpolation/CubicSpanInterpolator.hpp"
+#include "PointwiseFunctions/AnalyticSolutions/GeneralRelativity/KerrSchild.hpp"
+#include "Utilities/FileSystem.hpp"
+#include "tests/Unit/Evolution/Systems/Cce/BoundaryTestHelpers.hpp"
+#include "tests/Unit/TestCreation.hpp"
+
+SPECTRE_TEST_CASE("Unit.Evolution.Systems.Cce.OptionTags",
+                  "[Unit][Evolution]") {
+  CHECK(TestHelpers::test_creation<size_t, Cce::OptionTags::LMax>("8") == 8u);
```
Comment (kidder) [inline tests/Unit/Evolution/Systems/Cce/Test_OptionTags.cpp]: in `Utilities/Literals.hpp` there is code to make `8u` -> `8_st` for `size_t`, please use that

File: tests/Unit/Evolution/Systems/Cce/Test_OptionTags.cpp [hunk 34]
```diff
@@ -0,0 +1,55 @@
...
+            "0.5") == 0.5);
+  CHECK(TestHelpers::test_creation<std::string,
+                                   Cce::OptionTags::BoundaryDataFilename>(
+            "CceR0100.h5") == "CceR0100.h5");
+  CHECK(TestHelpers::test_creation<size_t, Cce::OptionTags::H5LookaheadTimes>(
+            "5") == 5u);
+  CHECK(TestHelpers::test_creation<size_t,
+                                   Cce::OptionTags::ScriInterpolationOrder>(
+            "4") == 4u);
+  const std::string filename = "CceR0100.h5";
```
Comment (kidder) [inline tests/Unit/Evolution/Systems/Cce/Test_OptionTags.cpp]: have the name of the file include the test name so we don't get conflicts when running tests in parallel

**Review by kidder — CHANGES_REQUESTED** [review]
joint with @nilsdeppe 

squash everything including the new comments; make sure the file add/deletion of `Test_InitializeCharacteristicExtractor` doesn't appear in any of the commits

**Comment by moxcodes** [comment]
Squashed.
Cheers!

**Review by nilsdeppe — APPROVED** [review]

**Comment by nilsdeppe** [comment]
clang-tidy failure is because of boost root finder internals warning that we can't fix

**Review by kidder — APPROVED** [review]

**Review by kidder — APPROVED** [review]

**Pull request merged**